### PR TITLE
hmem/synapseai: Refine the error handling and warning

### PR DESCRIPTION
**Pull request by shijin-aws**
Currently, synapseai_init return FI_ENODATA when there is no synapseai device on the platform. This will cause warnings are printed on any non-synapseai platform. This patch improves this by making synapseai_init return ENOSYS when all the API calls succeeded but zero devices are detected.

### CODE REVIEW

File: src/hmem_synapseai.c [hunk 1]
```diff
@@ -108,20 +108,31 @@ int synapseai_init(void)
...
 	status = synapseai_ops.synDeviceGetCount(&device_count);
-	if (status != synSuccess || device_count == 0)
-		/*
-		 * TODO We should call destroy here to free resources allocated
-		 * in initialize, but the destroy call hangs on instances without
-		 * a habana device
-		 */
-		return -FI_ENODATA;
+
+	/*
```
Comment (a-szegel) [inline src/hmem_synapseai.c:122]: We want to call synDestroy if `status != synSuccess` or `status != synSuccess`.  By separating these if statements, you make it more likely for us to miss this in the future.  I understand you did it b/c you want to return different error codes for each case, plz add a TODO comment for each block.

Comment (shijin-aws) [reply src/hmem_synapseai.c:122]: We should destroy unconditionally after synGetDeviceCount. That is the point. I can update the text to make it more clear

**Comment by j-xiong** [comment]
> when there is synapseai device on the platform

Is a `no` missing here?

**Comment by shijin-aws** [comment]
> > when there is synapseai device on the platform
> 
> Is a `no` missing here?

Yes, there is a typo, will update the commit

File: src/hmem_synapseai.c (hunk 1 above)
Comment (j-xiong) [reply src/hmem_synapseai.c:122]: Something like this should work: `Starting from here we should call destroy before returning error to free ....`

Comment (shijin-aws) [reply src/hmem_synapseai.c:122]: Updated

**Review by j-xiong — APPROVED** [review]

**Pull request merged**